# Лабораторная работа 4. Файлы и исключения
**Выполнил Мелкумян Сергей ИНБ-б-о-26-2**

## Задание

В файле data_v1.txt содержится последовательность символов, содержащая буквы латинского алфавита и десятичные цифры. Необходимо решить следующие задачи:
1. Найдите максимальное четное число в строке (под числом понимается последовательность цифр, ограниченная буквами).
2. Найдите буквенную подпоследовательность максимальной длины (то есть подпоследовательность не содержащую цифр). В ответе укажите ее длину.
3. Найдите символ в файле, который встречается чаще всего. Если таких символов несколько, то выведите все эти символы, в конце укажите одно число – сколько раз встречаются эти символы.


## Анализ задания, получение данных

Файл data_v1.txt расположен репозитории. Выполним копирование файла в локальное хранилище Google Colab, используя следующую команду:

In [ ]:
!wget https://raw.githubusercontent.com/enikolaev/IT-and-Programming/main/data/data_v1.txt


--2023-01-30 09:15:55--  https://raw.githubusercontent.com/enikolaev/IT-and-Programming/main/data/data_v1.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 100002 (98K) [text/plain]
Saving to: ‘data_v1.txt’

data_v1.txt         100%[===================>]  97.66K  --.-KB/s    in 0.002s  

2023-01-30 09:15:55 (47.5 MB/s) - ‘data_v1.txt’ saved [100002/100002]



Можно скопировать и поместить файл с данными любым удобным способом

Реализуем чтение всего файла в строку с использованием метоа `open()` и `read()`:

In [ ]:
f=open('data_v1.txt')
s=f.read().strip()
print(len(s))

100000


Обратите внимание на использование методв `strip()`

## Обработка символьной последовательности

Строки, подобно списка, являются коллекциями. Только в качестве элементов последовательности выступают символы. Для строк также используются функции нахождения длины `len()`, операция индексации `[]`

Например, можно посимвольно обходить строку, используя цикл `for` и индекс строки:

In [ ]:
# Вывод первых 10 символов строки с одновременным анализом буква это или цифра
for i in range(10):
  c = s[i]
  if c.isdigit():
    t = 'цифра'
  else:
    t='буква'
  print(c, t)

Z буква
S буква
t буква
M буква
P буква
i буква
7 цифра
c буква
s буква
i буква


## Решение задачи 1

Будем посимвольно обходимть строку и накапливать символы-цифры в строке `target`. Если цифровая последовательность прервется, то конвертируем `target` в число `int` и проверяем является ли число четным и не превысило ли оно максимум

In [ ]:
target=''
maxx=0
for c in s:
  if c.isdigit():
    target+=c
  else:
    if target!='':
      x = int(target)
      maxx=max(x, maxx)
      target=''

print('Максимальное четное число = %d' % maxx)


Максимальное четное число = 838478


## Решение задачи 2

Решение данной задачи осуществляется аналогично предыдущей. Можно даже вывести самуюдлинную буквенную строку

In [ ]:
target=''
maxx=0
maxs=''
for c in s:
  if c.isalpha():
    target+=c
  else:
    if len(target)>maxx:
      maxx=len(target)
      maxs=target
      target=''

print('Максимальная длина буквенной строки = %d' % maxx)
print('Сама строка:', maxs)

Максимальная длина буквенной строки = 1025
Сама строка: rpfunCMcMCfmnlNBbiCkCOPQsRYZmLjneFWZJjHqUxiTFGtMiKUTvaUkXBQnvGgpnDmUtFNyXZwWIdyqBWzLEqsTqoXCOkPaFQTuVxYeTftNLcEEjnRUqqAfeOmBHdcIDfyBpmvYjIivcZXMaflpfECMZKprJUSPIJtoolmBwWFmTiXtLsUgZrirYdtPUGEhsmAgGwqxoTcNlbtFmVApQAkcRiRFbvQxKwpdEVZhgUAymnZnRnJVmnDIvHmxtJkiICukqvSsMRrAHDBsAFDBWieGPhTcXNdetYzIdITVgvyiPqhPTKtixxggpYfCThpmcEqclpTTjnnHAXbpWOaOPlGcabgWvgiDDOhSWdTviUlNvTaoEVhoBfgrJkoJxtyDWXjVYYRokqkKPwRdkPkQYaVIoHCAfYPEyHgBsbUkYQyKNfgyWNJJsmtAoTPGrHKwUyJDvCEqgeRLTFdpqxaZFFGJrmkQuqbyadKxgWFsznrBtZyuubYwcmDAMqCqWXSUomuCMcFWWoxAWMJfdcYlDGMDSSCnYCUOVvgENGKZiTXKuLaLtMHTPjKmWwjwhfaePXiOPwZdvnAlPruuGwKhbRevjOHPVWIShguJJgKLaqUMJLwokvxJzYpvfOvfzRDHGQbnrWxESwCdsUzxyFWEYqRXpihePHiHXbMXBYyvIcUlaudLjXIPdXdMHrExnGgHeanzgtDSgSGOkxupzNQitVYWFXSNOpsdKMAveZQieNAqCMEXZzBdemldlsWCpiIMuQomlFEbZKcBwxAbNlFGPejoMQxgPHzXiskhadywliHoMmZlokYoauVlxmuHrqhwJHYpgzfUAqwPBAQPJNKlKhhYCRrLocFcIEorvZnBPIjlLVtMzafAoNJRufesTkbJcvviKEOSowgawXzWsFeUYgGsrdXZHbRGVQGvSczyAvdAeLjS

## Решение задачи 3

Для решения данной задачи будем использовать коллекцию словарь `dict()`. Данная коллекция содержит элемкнты в виде пар `ключ : значение`. Доступ к элементам осуществляется по ключу с использованием операции ндексации.

Заполним словарь d, который после прохода по строке должен принять вид:
d = {'A' : 10, 'B' : 127, '3' : 217, ... }
то есть в качестве ключа будем использовать символ, а вкачестве значения - количество таких символов в строке:

In [ ]:
# Создаем пустой словарь
d={}
# Проходим по строке и заполняем словарь
for c in s:
  d[c] = d.get(c, 0) + 1

# Выводим получившийся словарь:
print(d)

{'Z': 1623, 'S': 1613, 't': 1711, 'M': 1618, 'P': 1670, 'i': 1542, '7': 1643, 'c': 1562, 's': 1606, 'T': 1699, 'b': 1555, 'r': 1581, 'm': 1686, 'E': 1662, 'h': 1596, '8': 1654, 'f': 1609, 'B': 1647, 'K': 1616, 'a': 1586, 'F': 1623, 'O': 1580, 'w': 1582, 'o': 1609, '5': 1631, '1': 1580, 'W': 1595, '6': 1597, '0': 1651, 'I': 1604, 'u': 1606, 'e': 1566, 'p': 1644, 'D': 1644, 'A': 1638, 'Q': 1544, 'J': 1638, 'U': 1629, 'R': 1590, 'C': 1570, 'G': 1597, 'k': 1618, 'j': 1603, 'N': 1636, 'l': 1583, '2': 1600, 'q': 1614, 'X': 1619, 'H': 1565, '3': 1587, 'L': 1635, 'Y': 1647, '9': 1611, 'V': 1601, 'x': 1654, 'y': 1593, 'g': 1603, '4': 1599, 'v': 1655, 'd': 1599, 'z': 1615, 'n': 1566}


В получившемся словаре необходимо найти самый частотный символ. Для этого не будем обходить словарь, а просто воспользуемся встроенной функцией `max()`, которую применим к коллекции `values()` словаря

In [ ]:
max_val = max(d.values())
print(max_val)

1711


Выполним обход словаря и напечатаем ответ к данной задаче:

In [ ]:
for k in d:
  if d[k]==max_val:
    print(k, end=' ')
print(max_val)

t 1711


Для данного файла только один символ `t` встречается в файле максимальное `1711` количество раз

В данной работе были использованы операторы `if` (условный оператор), `for ... in range(...)` (оператор цикла с обходом диапазона). Применялись функции `print()`, `len()`, `max()`. В работе использовался тип данных `dict()`

## Индивидуальное задание. Вариант 9

В файле **data_v2.txt** найти самый частотный заглавный символ. В ответе указать сам символ и количество. Если таких символов несколько, указать все.

### Чтение файла

In [2]:
f = open('data_v2.txt', encoding='utf-8')
s = f.read().strip()
f.close()
print('Количество символов в файле:', len(s))

Количество символов в файле: 85775


### Поиск самого частотного заглавного символа

In [3]:
# Создаем словарь для подсчета заглавных символов
d = {}

for c in s:
  if c.isupper():
    d[c] = d.get(c, 0) + 1

max_val = max(d.values())

for k in d:
  if d[k] == max_val:
    print(k, end=' ')
print(max_val)

T 1452


**Ответ:** самый частотный заглавный символ — **T**, он встречается **1452** раз.

## Ответы на контрольные вопросы

**1. Опишите назначение, сходства и различия функций чтения из текстового файла: `read()`, `readline()`, `readlines()`.**

Все они читают данные из файла. `read()` считывает весь файл, `readline()` — одну строку, `readlines()` — все строки и возвращает их списком.

**2. Как прочитать многострочный текстовый файл, не используя явно методы `read()`, `readline()`, `readlines()`?**

Можно открыть файл и пройти по нему циклом `for`, получая строки по одной.

**3. Какова задача закрытия файла?**

Закрытие освобождает связанные с файлом ресурсы и завершает работу с ним.

**4. Что такое позиция считывания файла? Где первоначально находится позиция считывания при открытии файла?**

Это место в файле, с которого будет выполнено следующее чтение. При открытии файла позиция находится в начале файла.

**5. В каком режиме открывается файл, если требуется записать в него данные, но оставить существующее содержимое файла нетронутым? В какую часть файла данные записываются при этом?**

Файл открывается в режиме `a` (добавление). Новые данные записываются в конец файла.

**6. Дайте краткое объяснение, что такое исключение.**

Исключение — это ошибка или особая ситуация, возникающая во время выполнения программы.

**7. Что происходит, если вызвано исключение, и программа его не обрабатывает инструкцией `try/except`?**

Выполнение программы прекращается, и Python выводит сообщение об ошибке и traceback.

**8. Какое исключение программа вызывает, когда она пытается открыть несуществующий файл?**

`FileNotFoundError`.

**9. Какое исключение программа вызывает, когда она применяет функцию `float()` для конверсии нечелового строкового значения в число?**

`ValueError`.

**10. Что происходит, если файл не существует и программа пытается его открыть в режиме дозаписи?**

Файл будет создан. Если файл уже существует, новые данные будут добавляться в его конец.